src/middlewared/middlewared/main.py

```diff
@@ -1339,12 +1375,38 @@ def _call_prepare(
         audit_callback = audit_callback or (lambda message: None)
 
         args = []
-        if hasattr(methodobj, '_pass_app'):
+        has_pass_app = hasattr(methodobj, '_pass_app')
+        if has_pass_app:
             if methodobj._pass_app['require'] and app is None:
                 raise CallError('`app` is required')
 
             args.append(app)
 
+        ip = None
+        if hasattr(methodobj, '_no_auth_required'):
+            if not has_pass_app:
+                raise CallError(
+                    f'{serviceobj.__class__.__name__}.{methodobj.__name__} requires @pass_app decorator'
+                )
+            elif app is not None and isinstance(app.origin, TCPIPOrigin):
+                # it's important to rate limit based on remote addr/port
+                # combination. If we were to just apply this to the method
+                # then _ALL_ calls would be rate limited (including internal).
+                ip, port = app.origin.addr, app.origin.port
+                if not is_ha_connection(ip, port):
+                    key = '_'.join([name, ip])
```

Comment:
Should we also include port number in key?

Reply:
You could maybe just `key = '_'.join([name, str(app.origin)])` since the str method concatenates them already.